### ISSUE TITLE
persist: Kafka source permanently errored with "At least one input has no complete timestamps yet" if Mz killed within the --timestamp-frequency

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#9123]
### What version of Materialize are you using?

From the shell:

```
materialized v0.9.13-dev (9af55e568) 
```

### What was the issue?

If you kill Mz before the --timestamp-frequency tick has had the chance to happen, the source is permanently errored out after restart. 

Not reproducible without peristent kafka source. Not reproducible with low values of timestamp-frequency (naturally) or when persistent-user-tables is enabled(?)

### Is the issue reproducible? If so, please provide reproduction instructions.

1. Start Mz like this:
```
target/release/materialized --persistent-kafka-upsert-source --disable-persistent-system-tables-test  --experimental  --log-file=stderr 
```
2. Run
```
$ set-sql-timeout duration=5minutes
$ set keyschema={
    "type": "record",
    "name": "Key",
    "fields": [
        {"name": "f1", "type": "long"}
    ]
  }

$ set schema={
        "type" : "record",
        "name" : "test",
        "fields" : [
            {"name":"f1", "type":"long"}
        ]
    }

$ kafka-create-topic topic=s1

$ kafka-ingest format=avro topic=s1 key-format=avro key-schema=${keyschema} schema=${schema} publish=true repeat=100000
{"f1": ${kafka-ingest.iteration}} {"f1": ${kafka-ingest.iteration}}

> CREATE MATERIALIZED SOURCE s1
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-s1-${testdrive.seed}'
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE UPSERT

$ kafka-ingest format=avro topic=s1 key-format=avro key-schema=${keyschema} schema=${schema} publish=true repeat=100000
{"f1": ${kafka-ingest.iteration}} {"f1": ${kafka-ingest.iteration}}
$ kafka-ingest format=avro topic=s1 key-format=avro key-schema=${keyschema} schema=${schema} publish=true repeat=100000
{"f1": ${kafka-ingest.iteration}} {"f1": ${kafka-ingest.iteration}}
```
3. Immediately kil Mz and restart it
4. Source is permanently errored out:
```
materialize=> select count(*) from s1;
ERROR:  At least one input has no complete timestamps yet: [User(2)]
```

### Please attach any applicable log files.

If everything works, post-restart the Mz output has lines like this
```
2021-11-17T08:08:29.093189Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-s1-210 [0] (source u1/29) on worker 2    
2021-11-17T08:08:29.656995Z  INFO dataflow::source::kafka: Kafka message before expected offset, skipping: source kafka-u1/29 (reading topic testdrive-s1-210, partition 0) received offset 0 expected offset 300000    
2021-11-17T08:08:29.657089Z  WARN dataflow::source::kafka: Did not fast-forward consumer on partition PID: 0 to the correct Kafka offset. Currently at offset: 1 Expected offset: 300000    
```

If things do not work, the only line post-restart is:

```
2021-11-17T08:09:08.894105Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-s1-220 [0] (source u1/29) on worker 2    
```
and that is it.

**Comment by philip-stoev** [comment]
This is now reproducible inside mzcompose. It seems to require the following:
- the source must have been completely ingested prior to shutting down
- the time between complete ingestion and server shutdown should be less than the timestamp-frequency

To reproduce:

1. Clone `test-for-gh9123` from `git@github.com:philip-stoev/materialize.git`
2. Run
```
cd test/persistence
./mzcompose run kafka-sources
```

3. Test is set to wait for 5 min for the table to become available post-restart, but it never does:
```
> SELECT COUNT(*) = 100000 FROM s1;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 1s
kafka-sources/after-restart.td:1:1: error: executing query failed: db error: ERROR: At least one input has no complete timestamps yet: [User(2)]
     |
   1 | > SELECT COUNT(*) = 100000 FROM s1;
     | ^
```

**Comment by philip-stoev** [comment]
Same permanent error happens if writing to disk made impossible via a failpoint:

1. Run
```$ set keyschema={
    "type": "record",
    "name": "Key",
    "fields": [
        {"name": "f1", "type": "long"}
    ]
  }

$ set schema={
        "type" : "record",
        "name" : "test",
        "fields" : [
            {"name":"f2", "type":"long"}
        ]
    }

$ kafka-create-topic topic=fail-point

$ kafka-ingest format=avro topic=fail-point key-format=avro key-schema=${keyschema} schema=${schema} publish=true repeat=10000
{"f1": ${kafka-ingest.iteration}} {"f2": ${kafka-ingest.iteration}}

> CREATE MATERIALIZED SOURCE fail_point
  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-fail-point-${testdrive.seed}'
  WITH(topic_metadata_refresh_interval_ms=100)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}'
  ENVELOPE UPSERT

> SELECT COUNT(*) > 0 FROM fail_point;
true

$ kafka-ingest format=avro topic=fail-point key-format=avro key-schema=${keyschema} schema=${schema} publish=true start-iteration=10000 repeat=10000
{"f1": ${kafka-ingest.iteration}} {"f2": ${kafka-ingest.iteration}}

> SET failpoints = 'fileblob_set_sync=return';

$ kafka-ingest format=avro topic=fail-point key-format=avro key-schema=${keyschema} schema=${schema} publish=true start-iteration=20000 repeat=10000
{"f1": ${kafka-ingest.iteration}} {"f2": ${kafka-ingest.iteration}}
```
2. Kill and restart Mz
3. The `fail_point` source is permanently borked even though any other sources created prior to it are not errored out.

**Comment by aljoscha** [comment]
Thanks a lot! I think I figured this one out. It's not permanently wedged, you just have to wait long enough. The problem is in `filter_and_retract_future_updates()`. We have written out all of the `300_000` messages that the test ingests to persistence but need to retract them when starting up because they have not been _sealed_. I did not think that we would have huge loads of messages that need to be retracted so this method is horribly inefficient so far. It takes about 1s for reach retraction, meaning about 300_000 seconds, which is about 3.4 days. :sweat_smile: 

I'll change the operator to use the same batching and asynchronous waiting as the regular `persist()` operator, that should fix things.
